pkg/sql/revert.go

```diff
@@ -80,12 +90,127 @@ func RevertTables(
 		log.Infof(ctx, "reverting table %s (%d) to time %v", tables[i].GetName(), tables[i].GetID(), targetTime)
 	}
 
-	// TODO(dt): pre-split requests up using a rangedesc cache and run batches in
-	// parallel (since we're passing a key limit, distsender won't do its usual
-	// splitting/parallel sending to separate ranges).
-	for len(spans) != 0 {
-		var b kv.Batch
-		for _, span := range spans {
+	return RevertSpans(ctx, db, spans, targetTime, ignoreGCThreshold, batchSize, nil)
+}
+
+// RevertSpansContext provides the execution environment for a call to
+// RevertSpansFanout.
+type RevertSpansContext interface {
+	ExtendedEvalContext() *extendedEvalContext
+	ExecCfg() *ExecutorConfig
+	DistSQLPlanner() *DistSQLPlanner
+}
+
+// RevertSpansFanout calls RevertSpans in parallel. The span is
+// divided using DistSQL's PartitionSpans.
+//
+// We do this to get parallel execution of RevertRange even in the
+// case of a non-zero batch size. DistSender will not parallelize
+// requests with non-zero MaxSpanRequestKeys set.
+func RevertSpansFanout(
+	ctx context.Context,
+	db *kv.DB,
+	rsCtx RevertSpansContext,
+	spans []roachpb.Span,
+	targetTime hlc.Timestamp,
+	ignoreGCThreshold bool,
+	batchSize int64,
+	onCompletedCallback func(context.Context, roachpb.Span) error,
+) error {
+	ctx, sp := tracing.ChildSpan(ctx, "sql.RevertSpansFanout")
+	defer sp.Finish()
+
+	execCfg := rsCtx.ExecCfg()
+	maxWorkerCount := int(maxRevertSpanNumWorkers.Get(execCfg.SV()))
+	if maxWorkerCount == 1 {
+		return RevertSpans(ctx, db, spans, targetTime, ignoreGCThreshold, batchSize, onCompletedCallback)
+	}
+
+	dsp := rsCtx.DistSQLPlanner()
+	planCtx, _, err := dsp.SetupAllNodesPlanning(ctx, rsCtx.ExtendedEvalContext(), execCfg)
+	if err != nil {
+		return err
+	}
+
+	spanPartitions, err := dsp.PartitionSpans(ctx, planCtx, spans)
+	if err != nil {
+		return err
+	}
+
+	var workerPartitions []SpanPartition
```

Comment:
perhaps in a follow up PR: to maximize parallelism over time, we may want to distribute work to the workers over a queue as they work, instead of pre allocating work, the workers should merely pull the next range of work of some queue when they're done. Perhaps this isn't that big of a deal, if each range will have the same amount of work, but I'm not sure that's the case.

Reply:
Yeah, the reason I didn't use _just_ a queue here is part of the idea is that by dividing up the work with PartitionSpans, each worker has a list of spans that more or less goes to the same set of nodes, so we hopefully aren't sending a lot of RevertRange requests to the same node at once.

But we still still look at this for the future.